### ISSUE TITLE
Native bst support on Windows (not WSL)

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#1021]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/1021)
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on May 9, 2019, 13:54

Windows is a primary development platform for me, I want to be able to use BuildStream on Windows to build binaries for Windows.

Previously, we completed this ticket for remote-only support on WSL: #412

I aired this idea on the mailing list [here](https://mail.gnome.org/archives/buildstream-list/2019-April/msg00057.html), with a TODO list of primary concerns.

## Initial POC

Can compile using host tools, using a 'zero sandbox'.

* [x] !1298: Proof-of-concept of BuildStream building things on Windows.

## Rebased POC, after introduction of `buildbox-casd`, only supporting import elements

Only supports import elements, tested on `doc/examples/first-project/`.

* [x] Experimental branches for support buildbox-casd on Windows, working for 'import-only' workflow: BuildGrid/buildbox/buildbox-common!134, BuildGrid/buildbox/buildbox-casd!89.
* [x] Make a new minimal example of BuildStream working on Windows: https://gitlab.com/BuildStream/buildstream/commits/aevri/win32_minimal
* [x] #1154: Independently confirm the Windows experiment works.

## Official win32 support, using a 'zero sandbox'

Land the changes necessary to be able to run `doc/examples/first-project/` on Windows.

* [x] #1163, !1678: Add CI for native windows support.
* [ ] Full support for [`multiprocessing.set_start_method('spawn')`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.set_start_method), for platforms that don't support `fork`.
    * [ ] !1663: Enable `pytest tests/` with `BST_FORCE_START_METHOD=spawn`
    * [ ] !1675: Enable all tests for `BST_FORCE_START_METHOD=spawn`
    * [ ] Measure performance of spawn mode, determine if we should be mindful of what references what, so that we're not pickling
          more than we need to. This may mean being more explicit in general about the split between the main and worker processes.
* [ ] Change to use some more portable versions of things, re-assess some assumptions, being mindful of performance across platforms. Including:
    - Separators: consider `os.sep` **and** `os.altsep`
    - Signals, SIGINT, SIGTERM, SIGPIPE not on win32.
    - Use of add_child_handler to wait on subprocesses, not available on win32.
    - Use of `loop.add_reader(queue._reader.fileno() ...)` (maybe this works? `add_reader` and `_reader` seem to be available on Win32).
    - Takes a couple of seconds to spawn a BuildStream process on Windows.
    - Can't use UNIX domain sockets to communicate with buildbox-casd on Windows, grpc doesn't have that support, yet. 
    - ...
    - See !1633: WIP: win32: support 'bst show' with buildbox-casd and just enough signal support
* [ ] Support `CASCache.import_directory` in buildbox-casd on Windows.
* [ ] Fully support buildbox-casd on Windows.
* [ ] Introduce SandboxNone, which relies on copying files and good faith. I'd also
      like to try this on MacOS, to provide an alternative to the package manager
      Homebrew.
* [ ] Bonus: Provide a facility for plugin developers to record and replay job
      invocations, for debugging. This works by saving the pickled job data and
      launching it again separately.

## Proper sandboxing on Windows
- Evaluate [ProjFS](https://docs.microsoft.com/en-us/windows/desktop/projfs/projected-file-system) on Windows, to provide some filesystem sandboxing.
- Evaluate [Windows Sandbox](https://techcommunity.microsoft.com/t5/Windows-Kernel-Internals/Windows-Sandbox/ba-p/301849) for sandboxing.
- Evaluate [Windows Containers](https://docs.microsoft.com/en-us/virtualization/windowscontainers/about/) for sandboxing.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @barcharcraz]](https://gitlab.com/barcharcraz) on May 28, 2019, 18:57

For sandboxing I think you want Appcontainer Isolation (https://docs.microsoft.com/en-us/windows/desktop/SecAuthZ/appcontainer-for-legacy-applications-)

It won't do filesystem sand-boxing in the same way as linux namespaces but you can explicitly control what files / objects your spawned process has access to.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on May 29, 2019, 13:33

Will look at that too, thanks!

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Jul 17, 2019, 11:35

mentioned in merge request !1463

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Jul 17, 2019, 11:39

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Jul 17, 2019, 11:39

marked the task **!1298: Proof-of-concept hacking of BuildStream building things on Windows.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Jul 25, 2019, 10:42

marked the task **!1463: Make Jobs and the things they refer to picklable. (elements, sources, and more)** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Aug 1, 2019, 16:01

mentioned in merge request !1511

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Aug 1, 2019, 16:02

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Aug 20, 2019, 10:38

marked the task **!1511: Support [`multiprocessing.set_start_method('spawn')`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.set_start_method), for platforms that don't support `fork`.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Aug 23, 2019, 17:15

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 3, 2019, 16:49

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 3, 2019, 16:50

marked the task **Experimental branches for support buildbox-casd on Windows, working for 'import-only' workflow: BuildGrid/buildbox/buildbox-common!134, BuildGrid/buildbox/buildbox-casd!89.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 3, 2019, 17:28

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 3, 2019, 17:29

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 3, 2019, 17:31

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 3, 2019, 17:32

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Oct 7, 2019, 14:37

marked this issue as related to #1160

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on Oct 8, 2019, 14:30

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 8, 2019, 14:36

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 8, 2019, 14:54

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Oct 10, 2019, 14:38

Just something that I've spotted with some rebasing i'm doing. We've added an asyncio `get_child_watcher()` for monitoring the casd processes when executing the scheduler. This method is only compatible with Unix. (As you've already noted, we also use child_handlers for jobs)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 10, 2019, 15:08

Thanks for saying! Sounds like we'll need an abstraction over that, perhaps in `utils` or available from `platform`.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 17, 2019, 15:09

marked the task **!1624: Introduce win32 platform, using SandboxDummy.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 25, 2019, 10:30

marked the task **!1511: Support [`multiprocessing.set_start_method('spawn')`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.set_start_method), for platforms that don't support `fork`.** as incomplete

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 25, 2019, 10:41

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 25, 2019, 10:41

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 25, 2019, 10:41

marked the task **Use stable identifiers for `plugbase.PluginSource`s, so that we can support plugins that are more complicated than just an entrypoint class.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 25, 2019, 10:41

marked the task **Add an error if plugins are accidentally pickled, this makes debugging easier - otherwise we get cryptic namespace errors.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 25, 2019, 10:41

marked the task **Add an error if plugins are accidentally pickled, this makes debugging easier - otherwise we get cryptic namespace errors.** as incomplete

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 25, 2019, 10:41

marked the task **Make `git` available in the missing dependencies unit tests.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 25, 2019, 11:00

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 25, 2019, 11:24

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 25, 2019, 13:03

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 29, 2019, 15:07

marked the task **Make a new minimal example of BuildStream working on Windows: https://gitlab.com/BuildStream/buildstream/commits/aevri/win32_minimal** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 29, 2019, 15:09

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 29, 2019, 15:09

marked the task **!1654: Pickle first-pass factories, i.e. those used for junctions.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 29, 2019, 15:09

marked the task **!1511: Initial spawn support.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 31, 2019, 16:10

marked the task **!1657: Pickle node.pyx global state, reset on every bst invocation in tests.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 31, 2019, 16:19

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 31, 2019, 16:20

unassigned [[Gitlab user @aevri]](https://gitlab.com/aevri)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 31, 2019, 16:24

mentioned in merge request BuildGrid/buildbox/buildbox-common!134

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 31, 2019, 16:25

mentioned in merge request BuildGrid/buildbox/buildbox-casd!89

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 31, 2019, 16:29

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 31, 2019, 16:30

marked the task **#1163, !1678: Add CI for native windows support.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Oct 31, 2019, 16:31

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @aevri]](https://gitlab.com/aevri) on Nov 5, 2019, 12:17

marked the task **#1154: Independently confirm the Windows experiment works.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 8, 2020, 11:12

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 8, 2020, 12:27

mentioned in merge request !1633

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 8, 2020, 12:28

mentioned in merge request !1663

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 8, 2020, 12:28

mentioned in merge request !1675

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jun 15, 2020, 09:30

mentioned in merge request !1965

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jun 16, 2020, 17:22

As per [discussion on the ML](https://mail.gnome.org/archives/buildstream-list/2020-June/msg00013.html), support for pickling objects has been removed with MR !1965.

Plans include an investigation on removing the `fork()` model in favor of simple threading, and if that is not performant enough under the new landscape, to re-introduce pickling with subprocesses and asyncio (without multiprocessesing and `exec()`-less `fork()`).
